### PR TITLE
feat: add util func required by override

### PR DESCRIPTION
**Pull request by zhiying-lin**
### Description of your changes

- Add util func for rollout controller

Fixes #

I have:

- [X] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/controllers/rollout/override.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"k8s.io/klog/v2"
+
+	clusterv1beta1 "go.goms.io/fleet/apis/cluster/v1beta1"
+	fleetv1alpha1 "go.goms.io/fleet/apis/placement/v1alpha1"
+	fleetv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+	"go.goms.io/fleet/pkg/utils"
+	"go.goms.io/fleet/pkg/utils/controller"
+)
+
+func (r *Reconciler) getAllOverrides(ctx context.Context, crp string, masterResourceSnapshot *fleetv1beta1.ClusterResourceSnapshot) ([]*fleetv1alpha1.ClusterResourceOverride, []*fleetv1alpha1.ResourceOverride, error) {
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override.go:27]: nit; the function name is too vague, it's more of getAllMatchingOverridesForResourceSnapshot

File: pkg/controllers/rollout/override.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+				filteredRO = append(filteredRO, &roList.Items[i])
+				break
+			}
+		}
+	}
+	return filteredCRO, filteredRO, nil
+}
+
+// getLatestOverridesPerBinding will look for any overrides associated with the "Bound" binding.
+func (r *Reconciler) getLatestOverridesPerBinding(ctx context.Context, binding *fleetv1beta1.ClusterResourceBinding, croList []*fleetv1alpha1.ClusterResourceOverride, roList []*fleetv1alpha1.ResourceOverride) ([]string, []fleetv1beta1.NamespacedName, error) {
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override.go:119]: nit: the function name is more like pickFromResourceMatchedOverrideThatMatchesTargetCluster. Maybe add a few more comments?

File: pkg/controllers/rollout/override_test.go [hunk 3]
```diff
@@ -0,0 +1,905 @@
...
+					},
+					Spec: fleetv1alpha1.ClusterResourceOverrideSpec{
+						ClusterResourceSelectors: []fleetv1beta1.ClusterResourceSelector{
+							{
+								Group:   "",
+								Version: "v1",
+								Kind:    "Namespace",
+								Name:    "name-1",
+							},
+						},
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override_test.go:706]: does this field matter in this test?

File: pkg/controllers/rollout/override_test.go [hunk 4]
```diff
@@ -0,0 +1,905 @@
...
+							OverrideRules: []fleetv1alpha1.OverrideRule{
+								{
+									JSONPatchOverrides: []fleetv1alpha1.JSONPatchOverride{
+										{
+											Operator: fleetv1alpha1.JSONPatchOverrideOpAdd,
+											Path:     ".meta",
+											Value:    "abc",
+										},
+									},
+								},
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override_test.go:550]: why does this field matter in this test?

File: pkg/controllers/rollout/override_test.go [hunk 5]
```diff
@@ -0,0 +1,905 @@
...
+					},
+					Spec: fleetv1alpha1.ClusterResourceOverrideSpec{
+						ClusterResourceSelectors: []fleetv1beta1.ClusterResourceSelector{
+							{
+								Group:   "",
+								Version: "v1",
+								Kind:    "Namespace",
+								Name:    "name-1",
+							},
+						},
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override_test.go:539]: why does this field matter in this test?

File: pkg/controllers/rollout/override_test.go [hunk 6]
```diff
@@ -0,0 +1,905 @@
...
+			cluster: &clusterv1beta1.MemberCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-1",
+				},
+			},
+			wantCRO: []string{},
+			wantRO:  []fleetv1beta1.NamespacedName{},
+		},
+		{
+			name: "matched overrides with empty cluster label",
```
Comment (ryanzhang-oss) [inline pkg/controllers/rollout/override_test.go:520]: we should add a nagative RO/CRO in the list

File: pkg/controllers/clusterresourceplacement/resource_selector.go [hunk 7]
```diff
@@ -102,7 +102,7 @@ func (r *Reconciler) gatherSelectedResource(placement string, selectors []fleetv
 	return resources, nil
 }
 func sortResources(resources []runtime.Object) {
-	sort.Slice(resources, func(i, j int) bool {
+	sort.SliceStable(resources, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/clusterresourceplacement/resource_selector.go:105]: you're right, no possibility to have two identical resources, will revert

File: pkg/controllers/rollout/override_test.go [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+					},
+				},
+			},
+			croList: []fleetv1alpha1.ClusterResourceOverride{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "cro-1",
+					},
+					Spec: fleetv1alpha1.ClusterResourceOverrideSpec{
+						ClusterResourceSelectors: []fleetv1beta1.ClusterResourceSelector{
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/rollout/override_test.go:131]: you mean a matched CRO? the wantResult is empty.

The CRO itself is not valid as the policy is required. But this func does not care.

File: pkg/controllers/rollout/override_test.go [hunk 9]
```diff
@@ -0,0 +1,905 @@
...
+							{
+								Group:   "",
+								Version: "v1",
+								Kind:    "Namespace",
+								Name:    "svc-namespace",
```
[the remark this answers is not in the document]
Comment (zhiying-lin) [reply pkg/controllers/rollout/override_test.go:193]: yes, it creates a sample service used for testing.

File: pkg/controllers/rollout/override_test.go (hunk 4 above)
Comment (zhiying-lin) [reply pkg/controllers/rollout/override_test.go:550]: it does not. removed

File: pkg/controllers/rollout/override_test.go (hunk 5 above)
Comment (zhiying-lin) [reply pkg/controllers/rollout/override_test.go:539]: removed

File: pkg/controllers/rollout/override_test.go (hunk 3 above)
Comment (zhiying-lin) [reply pkg/controllers/rollout/override_test.go:706]: removed

**Review by ryanzhang-oss — APPROVED** [review]

**Pull request merged**